### DISCUSSION TITLE
SLF4JServiceProvider cannot be found

### DISCUSSION THREAD

**Discussion by Gregor7008** [qos-ch/slf4j#295] [Q&A] [answered]
Good evening everyone!
So I started trying to configure my own SLF4J implementation today, as I - as usual - wanted to do a bit more than [logback](https://github.com/qos-ch/logback) provides. That's also where I'm stuck now...
First of all: Am I correct in thinking, that by providing a SLF4JServiceProvider, a Logger and a respective LoggerFactory class I can reroute all SLF4J traffic to a GUI for example? If so....
For some reason my SLF4JServiceProvider implementation doesnt get found. I'm running SLF4J-API v2.0.0-beta1 and request 2.0.99 in the service provider. Just saying, in case that could be an issue.

As code reference I have pasted everything on pastebin:
- [SLF4JServiceProvider](https://pastebin.com/R7RqnZyC)
- [Logger](https://pastebin.com/Avbpy11S)
- [LoggerFactory](https://pastebin.com/nSEfE1pC)

Thank you for any help in advance - Have a great day!
Gregor

**Comment by ceki** [comment, marked as answer]
Have you placed a  provider-configuration file in the resource directory META-INF/services? That is, have you added a META-INF/services/org.slf4j.spi.SLF4JServiceProvider file containing the fully qualified class name of your provider?

See also https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html

**Reply by Gregor7008** [reply]
Problem solved - Thank you very much!

(I'm actually banging my head against a wall right now, because I went to exactly that doc entry to fix the problem by myself but somehow missed it... Anyways: It works now!)

**Comment by mattiasflodin** [comment]
Hi, the ServiceLoader stuff is not described anywhere in the FAQ under [How do I make my logging framework SLF4J compatible? ](https://www.slf4j.org/faq.html#slf4j_compatible). It tells you to use [slf4j-jdk14](https://github.com/qos-ch/slf4j/blob/master/slf4j-jdk14/src/main/java/org/slf4j/jul/JULServiceProvider.java) as a reference for how to implement a binding for your logging backend. But as far as I can tell, org.slf4j.spi.SLF4JServiceProvider which it inherits from does not exist.

Can you please point me to up-to-date documentation on how to implement a logger backend? If none exists, I'd like to request that the manual / FAQ is updated.

**Comment by EricHeath-AI** [comment]
If you are stuck with a compile time dependency against older slf4j versions, eg. a spring 5.x / spring boot combo (eg. https://github.com/spring-projects/spring-boot/blob/2.7.x/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/logback/LogbackLoggingSystem.java#L293) or similar, you could run a simple adapter like this:
```
package org.slf4j.impl;

import org.slf4j.ILoggerFactory;
import org.slf4j.LoggerFactory;
import org.slf4j.spi.LoggerFactoryBinder;

/**
 * <p>
 * This StaticLoggerBinder class is responsible for providing backwards
 * compatability with code that was compiled against older versions of slf4j.
 * </p>
 * <p>
 * The original StaticLoggerBinder had a different implementation, and was used
 * to bootstrap the startup of slf4j.  Some client code, ahem spring boot,
 * started relying on this class as a compile dependency.  Newer versions of
 * slf4j have removed this class entirely, which breaks libraries using it.
 * </p>
 * <p>
 * The newer slf4j just relies on LoggerFactory directly, and the standard java
 * Service provider / ServiceLoader mechanism detailed here,
 * https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.html,
 * to load logging implementations.  Since the existing StaticLoggerBinder
 * interface is small and seems compatible with the new pattern, this class
 * simply adapts the new behaviour to fit the old interface.  
 * </p>
 */
public class StaticLoggerBinder implements LoggerFactoryBinder {

    private static StaticLoggerBinder singleton = new StaticLoggerBinder();
    public static StaticLoggerBinder getSingleton() {
        return singleton;
    }
    
    public ILoggerFactory getLoggerFactory() {
        return LoggerFactory.getILoggerFactory();
    }

    public String getLoggerFactoryClassStr() {
        return getLoggerFactory().getClass().getName();
    }
}
```

It's not pretty, but you can upgrade slf4j and keep rolling...

**Comment by hazendaz** [comment]
@EricHeath-AI Now that that is tricky 👍 The other way with spring boot is to just disable their logging which is documented in the following https://github.com/spring-projects/spring-boot/issues/12649 and there are a couple of extra options in how to achieve that now too.  Kind of feel like above would be a smoother solution.

**Reply by EricHeath-AI** [reply]
Ah yes - I see their threads where people found `-Dorg.springframework.boot.logging.LoggingSystem=none` to be highly effective - thanks for that link.  In spring boot runtimes I agree with turning off their logging init if you are using old libraries or in the process of upgrading.  You end up hitting missing Joran classes as noted in the spring project comments as well.  So more patching than just what I listed above would be needed.

I did find good utility for the little adapter I posted when running a traditional web app / war housing, so I'll leave if here for other explorers.
